File: ruby/raytrace.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative 'tracer'
require_relative 'scene'

# ============================================================================
# TGA FILE OUTPUT
# ============================================================================

# Write image buffer as uncompressed 32-bit TGA file
# TGA format: simple raster format, no compression, BGRA byte order
def write_tga(filename, width, height, pixels)
  File.open(filename, 'wb') do |f|
    # TGA Header (18 bytes)
    f.write([0].pack('C'))          # ID length
    f.write([0].pack('C'))          # Color map type (0 = no color map)
    f.write([2].pack('C'))          # Image type (2 = uncompressed true-color)
    f.write([0].pack('v'))          # Color map first entry index
    f.write([0].pack('v'))          # Color map length
    f.write([0].pack('C'))          # Color map entry size
    f.write([0].pack('v'))          # X-origin
    f.write([0].pack('v'))          # Y-origin
    f.write([width].pack('v'))      # Image width
    f.write([height].pack('v'))     # Image height
    f.write([32].pack('C'))         # Pixel depth (32 bits = BGRA)
    f.write([8].pack('C'))          # Image descriptor (8 = origin at top-left, 8-bit alpha)

    # Write pixel data (BGRA format)
    # TGA expects bottom-to-top, but we write top-to-bottom with descriptor bit
    pixels.each do |color|
      # Convert float [0,1] to byte [0,255]
      r = (color.x * 255.0).clamp(0, 255).to_i
      g = (color.y * 255.0).clamp(0, 255).to_i
      b = (color.z * 255.0).clamp(0, 255).to_i
      a = 255

      # Write as BGRA
      f.write([b, g, r, a].pack('C4'))
    end
  end

  puts "Image written to #{filename}"
end

# ============================================================================
# CAMERA AND RENDERING
# ============================================================================

# Camera configuration
class Camera
  attr_accessor :position, :look_at, :up, :fov, :aspect

  def initialize
    @position = vec3(0, 0, 0)       # Camera position
    @look_at = vec3(0, 0, 1)        # Look direction
    @up = vec3(0, 1, 0)             # Up vector
    @fov = 60.0                     # Field of view in degrees
    @aspect = 16.0 / 9.0            # Aspect ratio
  end

  # Generate ray for pixel coordinates (u, v in [0, 1])
  def get_ray(u, v)
    # Calculate viewport dimensions
    theta = @fov * Math::PI / 180.0
    viewport_height = 2.0 * Math.tan(theta / 2.0)
    viewport_width = @aspect * viewport_height

    # Camera basis vectors
    w = (@position - @look_at).normalize  # Camera looks along -w
    u_vec = @up.cross(w).normalize        # Right vector
    v_vec = w.cross(u_vec)                # Corrected up vector

    # Viewport corners
    half_width = viewport_width / 2.0
    half_height = viewport_height / 2.0

    # Ray direction
    # Map from [0,1] to [-1,1] and scale by viewport dimensions
    horizontal = u_vec * viewport_width
    vertical = v_vec * viewport_height
    lower_left = @position - w - horizontal / 2.0 - vertical / 2.0

    # Generate ray
    direction = (lower_left + horizontal * u + vertical * v - @position).normalize
    Ray.new(@position, direction)
  end
end

# Main rendering function
def render(scene, width, height)
  pixels = []
  camera = Camera.new

  puts "Rendering #{width}x#{height} image..."
  puts "Camera at #{camera.position}, FOV #{camera.fov}°"
  puts "Scene contains #{scene.objects.length} objects"

  start_time = Time.now

  # Render each pixel
  height.times do |y|
    # Progress indicator
    if y % 20 == 0
      progress = (y.to_f / height * 100).to_i
      puts "Progress: #{progress}% (row #{y}/#{height})"
    end

    width.times do |x|
      # Calculate normalized coordinates [0, 1]
      u = x.to_f / (width - 1)
      v = 1.0 - (y.to_f / (height - 1)) # Flip Y for top-to-bottom raster

      # Generate ray
      ray = camera.get_ray(u, v)

      # Trace ray
      color = trace_ray(ray, scene)

      pixels << color
    end
  end

  elapsed = Time.now - start_time
  puts "Rendering complete in #{elapsed.round(2)} seconds"

  pixels
end

# ============================================================================
# MAIN PROGRAM
# ============================================================================

def main
  puts '=' * 60
  puts 'Ruby Educational Raytracer'
  puts 'Rendering scene with 5 primitive types'
  puts '=' * 60
  puts

  # Setup scene
  scene = setup_scene

  # Render configuration
  # Start with small resolution for testing
  # Students can increase to 1920x1080 for final renders
  width = 640
  height = 360

  # Render
  pixels = render(scene, width, height)

  # Write output
  write_tga('output.tga', width, height, pixels)

  puts
  puts 'Done! Open output.tga to view the rendered image.'
  puts
end

# Run the raytracer
main if __FILE__ == $PROGRAM_NAME

File: prolog/raytrace.pl
% raytrace.pl
% Main entry point: handles image buffer, camera setup, and TGA output

:- ['scene.pl'].
:- initialization(main, main).

% =============================================================================
% IMAGE CONFIGURATION
% =============================================================================

% Image dimensions (smaller for faster testing)
image_width(400).
image_height(300).

% Camera parameters
camera_fov(60.0).  % Field of view in degrees

% =============================================================================
% CAMERA RAY GENERATION
% =============================================================================

% Generate a ray for pixel (X, Y)
generate_ray(X, Y, Width, Height, FOV, Ray) :-
    % Convert pixel coordinates to normalized device coordinates [-1, 1]
    AspectRatio is Width / Height,
    HalfWidth is Width / 2.0,
    HalfHeight is Height / 2.0,
    NDC_X is (X - HalfWidth) / HalfWidth,
    NDC_Y is -(Y - HalfHeight) / HalfHeight,  % Flip Y (image origin is top-left)

    % Convert to camera space
    FOV_Rad is FOV * pi / 180.0,
    HalfFOV is FOV_Rad / 2.0,
    ViewportHeight is 2.0 * tan(HalfFOV),
    ViewportWidth is ViewportHeight * AspectRatio,

    CamX is NDC_X * ViewportWidth / 2.0,
    CamY is NDC_Y * ViewportHeight / 2.0,
    CamZ is 1.0,  % Look along +Z axis

    % Create ray
    Origin = vec3(0, 0, 0),
    Direction_unnorm = vec3(CamX, CamY, CamZ),
    vec3_normalize(Direction_unnorm, Direction),
    Ray = ray(Origin, Direction).

% =============================================================================
% COLOR CONVERSION
% =============================================================================

% Convert floating-point color [0,1] to byte [0,255]
color_to_byte(F, B) :-
    clamp(F, 0.0, 1.0, Clamped),
    % Apply simple gamma correction (gamma = 2.2)
    Gamma is Clamped ** (1.0 / 2.2),
    B is floor(Gamma * 255.0).

% =============================================================================
% RENDERING LOOP
% =============================================================================

% Render the entire image
render_image(Scene, Width, Height, FOV, Pixels) :-
    render_rows(Scene, Width, Height, FOV, 0, Height, [], Pixels).

% Render all rows
render_rows(_, _, Height, _, Height, Height, Acc, Pixels) :-
    reverse(Acc, Pixels).

render_rows(Scene, Width, Height, FOV, Y, MaxY, Acc, Pixels) :-
    Y < MaxY,
    render_row(Scene, Width, Height, FOV, 0, Width, Y, [], RowPixels),
    NextY is Y + 1,
    append(Acc, RowPixels, NewAcc),
    render_rows(Scene, Width, Height, FOV, NextY, MaxY, NewAcc, Pixels).

% Render a single row
render_row(_, Width, _, _, Width, Width, _, Acc, RowPixels) :-
    reverse(Acc, RowPixels).

render_row(Scene, Width, Height, FOV, X, MaxX, Y, Acc, RowPixels) :-
    X < MaxX,
    generate_ray(X, Y, Width, Height, FOV, Ray),
    trace_ray(Ray, Scene, Color),
    Color = vec3(R, G, B),
    color_to_byte(R, BR),
    color_to_byte(G, BG),
    color_to_byte(B, BB),
    Pixel = pixel(BR, BG, BB),
    NextX is X + 1,
    render_row(Scene, Width, Height, FOV, NextX, MaxX, Y, [Pixel|Acc], RowPixels).

% =============================================================================
% TGA FILE OUTPUT
% =============================================================================

% Write TGA header
write_tga_header(Stream, Width, Height) :-
    % TGA header (18 bytes)
    put_byte(Stream, 0),    % ID length
    put_byte(Stream, 0),    % Color map type
    put_byte(Stream, 2),    % Image type (2 = uncompressed true-color)

    % Color map specification (5 bytes, all zeros)
    put_byte(Stream, 0), put_byte(Stream, 0),
    put_byte(Stream, 0), put_byte(Stream, 0),
    put_byte(Stream, 0),

    % Image specification
    put_byte(Stream, 0), put_byte(Stream, 0),  % X origin
    put_byte(Stream, 0), put_byte(Stream, 0),  % Y origin

    % Width (little-endian)
    WidthLow is Width mod 256,
    WidthHigh is Width // 256,
    put_byte(Stream, WidthLow),
    put_byte(Stream, WidthHigh),

    % Height (little-endian)
    HeightLow is Height mod 256,
    HeightHigh is Height // 256,
    put_byte(Stream, HeightLow),
    put_byte(Stream, HeightHigh),

    put_byte(Stream, 24),   % Bits per pixel (24 = RGB)
    put_byte(Stream, 0).    % Image descriptor

% Write pixel data (BGR format for TGA)
write_pixels(_, []).
write_pixels(Stream, [pixel(R, G, B)|Rest]) :-
    put_byte(Stream, B),  % Blue
    put_byte(Stream, G),  % Green
    put_byte(Stream, R),  % Red
    write_pixels(Stream, Rest).

% Write complete TGA file
write_tga(Filename, Width, Height, Pixels) :-
    open(Filename, write, Stream, [type(binary)]),
    write_tga_header(Stream, Width, Height),
    write_pixels(Stream, Pixels),
    close(Stream).

% =============================================================================
% MAIN ENTRY POINT
% =============================================================================

main :-
    write('Starting raytracer...'), nl,

    % Get image dimensions
    image_width(Width),
    image_height(Height),
    camera_fov(FOV),

    write('Image size: '), write(Width), write('x'), write(Height), nl,
    write('Setting up scene...'), nl,

    % Setup scene
    setup_scene(Scene),

    write('Rendering...'), nl,

    % Render image
    statistics(walltime, [Start|_]),
    render_image(Scene, Width, Height, FOV, Pixels),
    statistics(walltime, [End|_]),
    RenderTime is End - Start,

    write('Render time: '), write(RenderTime), write(' ms'), nl,
    write('Writing output.tga...'), nl,

    % Write TGA file
    write_tga('output.tga', Width, Height, Pixels),

    write('Done!'), nl,
    halt.

main :-
    write('Error: Rendering failed'), nl,
    halt(1).
